lib/Dialect/TTIR/Analysis/DFShardingPolicy.cpp

```diff
@@ -0,0 +1,172 @@
+        currentOp = nullptr;
+        continue;
+      }
+
+      if (scheduler.hasUnscheduledOps()) {
+        scheduleableOps = scheduler.getScheduleableOps();
+
+        // Check if currentOp has a valid successor.
+        //
+        Operation *nextOp = nullptr;
+        for (auto op : scheduleableOps) {
+          for (auto operand : op->getOperands()) {
+            if (operand.getDefiningOp() == currentOp) {
+              nextOp = op;
+              break;
+            }
+          }
+        }
+
+        if (nextOp) {
+
+          // V0 Before adding to shard chain config check that currentOp is not
+          // fork/join op.
+          //
+          bool validForSharding = currentOp->hasOneUse() &&
+                                  legalGrids.lookup(currentOp).size() > 0 &&
+                                  legalGrids.lookup(nextOp).size() > 0;
+
+          // TODO
+          // It seems that bunch of TTNN ops have constraints which prevent
+          // them from being sharded if both inputs are interlieved, so my
+          // proposal is to try to resolve this by starting a shard chain with
+          // reshard op(at later phase only when necessary)
+          //
+
+          if (validForSharding) {
+            // Fetch largest legal sharded L1 grids for currentOp and nextOp.
+            //
+            // Calculate L1 tensor memory usage based on :
+            // currentOp output tensor shard spec, nextOp exec and nextOp output
+            // tensor.
+            //
+            LayoutAttr currentOpLayout = legalGrids.lookup(currentOp).front();
+            llvm::SmallVector<int64_t> currentOpShardShape =
+                currentOpLayout.getShardShape(false /*convertTileToScalar*/);
+            uint64_t l1InputUsage = currentOpLayout.getElementSizeBytes();
+            for (int64_t dim : currentOpShardShape) {
+              l1InputUsage *= dim;
+            }
+
+            LayoutAttr nextOpLayout = legalGrids.lookup(nextOp).front();
+            llvm::SmallVector<int64_t> nextOpShardShape =
+                nextOpLayout.getShardShape(false /*convertTileToScalar*/);
+            uint64_t l1OutputUsage = nextOpLayout.getElementSizeBytes();
+            for (int64_t dim : nextOpShardShape) {
+              l1OutputUsage *= dim;
+            }
+
+            // Figure out this const based on exec data, but will be replaced
+            // with API.
+            //
+            constexpr float tensorL1UsageCap = 0.8;
+            bool l1UsageValid = (l1InputUsage + l1OutputUsage) <
+                                tensorL1UsageCap * usableL1CacheSize;
+
+            if (l1UsageValid) {
+              // Add to shard chain config.
+              //
+              ShardSpec shardSpec;
+              shardSpec.op = currentOp;
+
+              // Hardcoded tensor split factor for now, until pipeline OP
+              // support is added.
+              //
+              shardSpec.tensorSplitFactor = 1;
```

Comment:
for my education what is tensorSplitFactor?

Reply:
When tensors cant fit in L1 we can create shard spec with some splitting factor. We can then try to reconcile whole or few parts of shard chain to same splitting factor. This will allow us to shard but with usage of pipeline op.(it is explained in Optimizer design doc)